### PR TITLE
Dennok's old shuttle decal fix

### PR DESCRIPTION
**Pull request by tralezab**
## About The Pull Request

Shuttle decals will work and this won't break bespoke

## Why It's Good For The Game

Fix

## Changelog
:cl:
fix: shuttle decals will work.
/:cl:

### CODE REVIEW

**Comment by TemporalOroboros** [comment]
Does this also apply to smoothing decals? If I remember correctly they copy some normal decal code because of needing to pass extra args when the decal changes.

**Comment by tralezab** [comment]
maybe 🤡

**Comment by tralezab** [comment]
Ok, implemented anturk's mostly completed fix, and it works. Looks like this is good to go

**Comment by tralezab** [comment]
really was a draft for only a couple seconds huh

**Comment by cacogen** [comment]
>doesn't say bugs is bad

one job

**Comment by tralezab** [comment]
bro...

File: code/__DEFINES/dcs/signals.dm [hunk 1]
```diff
@@ -284,6 +284,8 @@
 #define COMSIG_TURF_MULTIZ_DEL "turf_multiz_del"
 ///from base of turf/multiz_turf_new: (turf/source, direction)
 #define COMSIG_TURF_MULTIZ_NEW "turf_multiz_new"
+///from base of turf/proc/onShuttleMove(): (turf/newT) //E
```
Comment (Jared-Fogle) [inline code/__DEFINES/dcs/signals.dm:287]: //E?

File: code/datums/elements/decals/_decal.dm [hunk 2]
```diff
@@ -3,14 +3,60 @@
...
+	var/list/resulting_decals_params = list() // param lists
+	var/list/old_decals = list() //instances
+
+	if(!source.comp_lookup || !source.comp_lookup[COMSIG_ATOM_UPDATE_OVERLAYS])
+		//should probably also unregister itself
+		return
+
+	switch(length(source.comp_lookup[COMSIG_ATOM_UPDATE_OVERLAYS]))
+		if(0)
+			var/datum/element/decal/D = source.comp_lookup[COMSIG_ATOM_UPDATE_OVERLAYS]
```
Comment (Jared-Fogle) [inline code/datums/elements/decals/_decal.dm:25]: Better variable names.

File: code/datums/elements/decals/_decal.dm [hunk 3]
```diff
@@ -3,14 +3,60 @@
...
+			if(!istype(D))
+				return
+			old_decals += D
+			resulting_decals_params += list(D.get_rotated_parameters(old_dir,new_dir))
+		else
```
Comment (Jared-Fogle) [inline code/datums/elements/decals/_decal.dm:30]: Just use an if statement, not a switch.

File: code/datums/elements/decals/_decal.dm [hunk 4]
```diff
@@ -3,14 +3,60 @@
...
+	if(old_dir == new_dir)
+		return
+	var/list/resulting_decals_params = list() // param lists
+	var/list/old_decals = list() //instances
+
+	if(!source.comp_lookup || !source.comp_lookup[COMSIG_ATOM_UPDATE_OVERLAYS])
+		//should probably also unregister itself
+		return
+
+	switch(length(source.comp_lookup[COMSIG_ATOM_UPDATE_OVERLAYS]))
```
Comment (Jared-Fogle) [inline code/datums/elements/decals/_decal.dm:23]: What is this for? Are you sure this is sane? It's scary to me to be using internal stuff like this.

File: code/datums/elements/decals/_decal.dm [hunk 5]
```diff
@@ -3,14 +3,60 @@
...
+				return
+			old_decals += D
+			resulting_decals_params += list(D.get_rotated_parameters(old_dir,new_dir))
+		else
+			for(var/datum/element/decal/D in source.comp_lookup[COMSIG_ATOM_UPDATE_OVERLAYS])
+				old_decals += D
+				resulting_decals_params += list(D.get_rotated_parameters(old_dir,new_dir))
+
+	//Instead we could generate ids and only remove duplicates to save on churn on four-corners symmetry ?
+	for(var/datum/element/decal/D in old_decals)
```
Comment (Jared-Fogle) [inline code/datums/elements/decals/_decal.dm:36]: Better variable names.

File: code/datums/elements/decals/_decal.dm [hunk 6]
```diff
@@ -3,14 +3,60 @@
...
+		D.Detach(source)
+
+	for(var/result in resulting_decals_params)
+		source._AddElement(result)
```
Comment (Jared-Fogle) [inline code/datums/elements/decals/_decal.dm:40]: Don't use the internal `_AddElement`. Use `AddElement`.

Comment (Jared-Fogle) [reply code/datums/elements/decals/_decal.dm:40]: I know *why* you're doing this, but you shouldn't. This is an implementation detail.

File: code/datums/elements/decals/_decal.dm [hunk 7]
```diff
@@ -77,3 +116,11 @@
...
 
 	examine_list += description
+
+/datum/element/decal/proc/shuttle_move_react(datum/source, turf/newT)
+	SIGNAL_HANDLER
+
+	if(newT == source)
+		return
+	Detach(source)
+	newT.AddElement(/datum/element/decal, pic.icon, pic.icon_state, directional , cleanable, pic.color, pic.layer, pic.plane, description, pic.alpha)
```
Comment (Jared-Fogle) [inline code/datums/elements/decals/_decal.dm:126]:
```suggestion
	newT.AddElement(/datum/element/decal, pic.icon, pic.icon_state, directional, cleanable, pic.color, pic.layer, pic.plane, description, pic.alpha)
```

File: code/modules/shuttle/on_move.dm [hunk 8]
```diff
@@ -64,6 +64,8 @@ All ShuttleMove procs go here
 	if(isopenturf(newT))
 		var/turf/open/new_open = newT
 		new_open.copy_air_with_tile(src)
+	//Decal stuff
```
Comment (Jared-Fogle) [inline code/modules/shuttle/on_move.dm:67]:
```suggestion
```

Useless comment.

File: code/datums/elements/decals/_decal.dm [hunk 9]
```diff
@@ -77,3 +116,11 @@
 	SIGNAL_HANDLER
 
 	examine_list += description
+
+/datum/element/decal/proc/shuttle_move_react(datum/source, turf/newT)
```
Comment (Jared-Fogle) [inline code/datums/elements/decals/_decal.dm:120]: `new_turf`

File: code/datums/elements/decals/_decal.dm [hunk 10]
```diff
@@ -3,14 +3,60 @@
...
+	if(directional) //Even when the dirs are the same rotation is coming out as not 0 for some reason
+		rotation = SIMPLIFY_DEGREES(dir2angle(new_dir)-dir2angle(old_dir))
+		new_dir = turn(pic.dir,-rotation)
+	return list(/datum/element/decal, pic.icon, pic.icon_state, new_dir, cleanable, pic.color, pic.layer, pic.plane, description, pic.alpha)
+
+
+/datum/element/decal/Attach(atom/target, _icon, _icon_state, _dir, _cleanable=FALSE, _color, _layer=TURF_LAYER, _plane=FLOOR_PLANE, _description, _alpha=255, mutable_appearance/_pic)
 	. = ..()
-	if(!isatom(target) || !generate_appearance(_icon, _icon_state, _dir, _layer, _plane, _color, _alpha, target))
+	if(!isatom(target) || ( !_pic && !generate_appearance(_icon, _icon_state, _dir, _layer, _plane, _color, _alpha, target)))
```
Comment (Jared-Fogle) [inline code/datums/elements/decals/_decal.dm:53]:
```suggestion
	if(!isatom(target) || (!_pic && !generate_appearance(_icon, _icon_state, _dir, _layer, _plane, _color, _alpha, target)))
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]

**Pull request closed without merging**